### PR TITLE
Rebuild on mid build file modification

### LINKED ISSUE #2426: cargo build does not rebuild if a source file was modified during a build
I haven't checked, but I would guess this is a timestamp issue: the input is modified before the output artifact is created, even though the modified input was not used to build the artifact. This would happen any time after rustc is parsing.

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes #2426 

---

_original_

Trying to create a test for #2426, as described by @alexcrichton in https://github.com/rust-lang/cargo/issues/2426#issuecomment-382020985.

Alex, what am I doing wrong? Why isn't my test failing as it should be?

?r @alexcrichton

### CODE REVIEW

File: tests/testsuite/build.rs [hunk 1]
```diff
@@ -2809,6 +2811,112 @@ fn rebuild_preserves_out_dir() {
...
+            &format!(
+                r#"
+                extern crate proc_macro;
+
+                use std::net::TcpStream;
+                use proc_macro::TokenStream;
+
+                #[proc_macro_derive(Noop)]
+                pub fn noop(_input: TokenStream) -> TokenStream {{
+                    drop(TcpStream::connect("{}").unwrap());
```
Comment (alexcrichton) [inline tests/testsuite/build.rs]: After you `connect` here you'll want to `read_to_end` to block waiting for the disconnect to happen

File: tests/testsuite/build.rs [hunk 2]
```diff
@@ -2809,6 +2811,112 @@ fn rebuild_preserves_out_dir() {
...
+                }}
+            "#,
+                addr
+            ),
+        )
+        .build();
+    let root = p.root();
+
+    let t = thread::spawn(move || {
+        drop(server.accept().unwrap());
```
Comment (alexcrichton) [inline tests/testsuite/build.rs]: You'll want to defer dropping this socket until after the file has been modified

**Comment by dwijnand** [comment]
Still passes :(

**Comment by alexcrichton** [comment]
Hm it's interesting that the test failed on Windows but succeeded on Unix? I wonder if the file needs to be closed? This may also have to do with weird trickery around timestamp resolution. I'll take a closer look tomorrow at Cargo's code here to try to pinpoint the bug

**Comment by ehuss** [comment]
If you add `drop(file);` just before `drop(socket);`, then it reliably passes on my Windows system.  I suspect it has to do with threading differences between the platforms.

This test fails about half the time on my Mac.  It's because my filesystem only has 1s resolution.  So if the fingerprint is saved at 1234.1 and you modify the file at time 1234.2, the file is considered fresh because 1234 == 1234.

To make this reliably fail on all platforms, you want to modify `root/src/lib.rs`, not the `proc-macro`.   Modifying the proc-macro is modifying the thing that was just built, not the thing that is currently building.

**Comment by dwijnand** [comment]
You're right, @ehuss. Thank you. This is now consistently failing on me. 🎉

**Comment by dwijnand** [comment]
I have a fix for my test.

Please review - I'm probably doing something silly (design wise, or just Rust-wise).

**Comment by alexcrichton** [comment]
Hm it looks like some other tests may have started failing? I think the main fix here is that we'll basically want to record the mtime of the start of the build rather than the end of the build, and I think that should fix issue where files are modified during the build?

**Comment by dwijnand** [comment]
Do you mean by changing `LocalFingerprint`'s structure to not reference a path?

**Comment by alexcrichton** [comment]
Something like that yeah. The current mtime we base everything off is the dep-info mtime which the compiler emits (I think?), but the relevant mtime is rather when the build started.

**Comment by alexcrichton** [comment]
@dwijnand that's what's called after the build is completed and it just updates with the mtime of the path mentioned (as the path may not have existed before the build started). I think that method would basically go away and we'd just record what time the build started, then we'd record that.

**Comment by dwijnand** [comment]
@alexcrichton how do I avoid spurious recompiles? I have an implementation that just uses the build start time, but it breaks my test because now the second compile recompiles `proc_macro_dep` when it needn't.

**Comment by alexcrichton** [comment]
Hm I'm not entirely sure! Want to set `RUST_LOG=cargo::core::compiler::fingerprint` and see why Cargo is recompiling?

**Comment by dwijnand** [comment]
It's recompiling because the fingerprints, which now hold the build start time, don't match..

```
running `/d/cargo/target/debug/cargo build`
running `/d/cargo/target/debug/cargo build`
DEBUG 2018-04-27T14:33:22Z: cargo::core::compiler::fingerprint: fingerprint at: /d/cargo/target/cit/t0/p/target/debug/.fingerprint/root-819987bee50907ec/lib-root-819987bee50907ec
 INFO 2018-04-27T14:33:22Z: cargo::core::compiler::fingerprint: fingerprint error for root v0.1.0 (file:///d/cargo/target/cit/t0/p/root): mtime based components have changed: previously Some(FileTime { seconds: 1524839602, nanos: 146249000 }) now Some(FileTime { seconds: 1524839602, nanos: 283090000 })
DEBUG 2018-04-27T14:33:22Z: cargo::core::compiler::fingerprint: fingerprint at: /d/cargo/target/cit/t0/p/target/debug/.fingerprint/proc_macro_dep-aae81e26f117d294/lib-proc_macro_dep-aae81e26f117d294
 INFO 2018-04-27T14:33:22Z: cargo::core::compiler::fingerprint: fingerprint error for proc_macro_dep v0.1.0 (file:///d/cargo/target/cit/t0/p/proc_macro_dep): mtime based components have changed: previously Some(FileTime { seconds: 1524839601, nanos: 867084000 }) now Some(FileTime { seconds: 1524839602, nanos: 282926000 })
   Compiling proc_macro_dep v0.1.0 (file:///d/cargo/target/cit/t0/p/proc_macro_dep)
DEBUG 2018-04-27T14:33:23Z: cargo::core::compiler::fingerprint: write fingerprint: /d/cargo/target/cit/t0/p/target/debug/.fingerprint/proc_macro_dep-aae81e26f117d294/lib-proc_macro_dep-aae81e26f117d294
   Compiling root v0.1.0 (file:///d/cargo/target/cit/t0/p/root)
DEBUG 2018-04-27T14:33:23Z: cargo::core::compiler::fingerprint: write fingerprint: /d/cargo/target/cit/t0/p/target/debug/.fingerprint/root-819987bee50907ec/lib-root-819987bee50907ec
    Finished dev [unoptimized + debuginfo] target(s) in 1.13 secs
```

**Comment by alexcrichton** [comment]
Hm sorry I'm not immediately sure what's going on. The fingerprint module is pretty finnicky and fixing this issue may be exposing existing bugs that otherwise need to be explored :(

**Comment by alexcrichton** [comment]
I believe (trying to remember right) the way things work today is something like:

* If no output artifacts exist, we do a compilation.
* Otherwise we assume that the dep-info file (`*.d`) was an output artifact. We also assume that all files necessary for the compilation are listed.
* The mtime of all files listed previously are compared to the mtime of the `*.d` file, if any are newer then we recompile
* After a crate finishes compiling we update the `*.d` with Cargo's own internal format/serialization.

The last step there means that when we update the `*.d` file it's *after* the compilation so any modifications that happened during the compilation aren't tracked because they're considered older than the mtime of the `*.d` file, thend of compilation.

I think the fix here will basically be to tweak [`LocalFingerprint::MtimeBased`](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/compiler/fingerprint.rs#L198-L203) slightly. Ideally when `update_local` is called it *doesn't* call `mtime` because that time is too far in the future. Rather it should update the mtime of the file listed (if it's a Cargo-internal file) to just before compilation started (perhaps an extra payload to the `MtimeBased` variant)

**Comment by dwijnand** [comment]
@alexcrichton I had a go implementing what you said (or at least the parts I understood and knew how to do). It's still failing my test, but I added some debug info to the commit message too. If you have time to review that would be great! 🙇

**Comment by dwijnand** [comment]
Actually I think I'm done. It was too difficult for me, at this time.

**Comment by dwijnand** [comment]
Thanks @alexcrichton. It was fun to learn to about proc macros and using TCP to test this behaviour. I guess that effort could be reused by the next person attacking #2426.

Actually: does Cargo's test suite have "pending until fixed" test facilities? By that I mean you can assert that a test doesn't pass because it's blocked on issue X being fixed? If so perhaps I could land my test in master.

Finally: If you keep labeling things E-easy/E-help-wanted/E-mentor I'll keep an eye out for them. 😄

**Comment by alexcrichton** [comment]
I think we may have a test here or there for unfixed bugs but in-the-large we're only checking in working tests

**Pull request closed without merging**